File: IMEIndicator/Models/ProcessPriorityRule.cs
using System.Diagnostics;
using System.Text.Json.Serialization;

namespace IMEIndicator.Models;

/// <summary>
/// Windowsプロセス優先度（タスクマネージャー準拠の6段階）
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum PriorityLevel
{
    /// <summary>低</summary>
    Idle,
    /// <summary>通常以下</summary>
    BelowNormal,
    /// <summary>通常</summary>
    Normal,
    /// <summary>通常以上</summary>
    AboveNormal,
    /// <summary>高</summary>
    High,
    /// <summary>リアルタイム</summary>
    Realtime
}

/// <summary>
/// プロセス優先度制御ルール
/// </summary>
public class ProcessPriorityRule
{
    /// <summary>
    /// 監視対象プロセス名（.exe拡張子あり/なし両対応）
    /// </summary>
    public string ProcessName { get; set; } = string.Empty;

    /// <summary>
    /// 設定する優先度
    /// </summary>
    public PriorityLevel TargetPriority { get; set; } = PriorityLevel.Normal;

    /// <summary>
    /// バックオフ最大指数。最大間隔 = PollingIntervalSeconds * 2^MaxBackoffExponent
    /// </summary>
    public int MaxBackoffExponent { get; set; } = 6;

    /// <summary>
    /// ルールの有効/無効
    /// </summary>
    public bool IsEnabled { get; set; } = true;

    /// <summary>
    /// E-Core（効率コア）のみに固定するかどうか
    /// </summary>
    public bool UseECoreOnly { get; set; } = false;

    /// <summary>
    /// バリデーション済みの MaxBackoffExponent を返す（0～10）
    /// </summary>
    [JsonIgnore]
    public int ValidatedMaxBackoffExponent => Math.Clamp(MaxBackoffExponent, 0, 10);

    /// <summary>
    /// .exe 拡張子を除去したプロセス名を返す
    /// </summary>
    [JsonIgnore]
    public string NormalizedProcessName
    {
        get
        {
            var name = ProcessName.Trim();
            if (name.EndsWith(".exe", StringComparison.OrdinalIgnoreCase))
                name = name[..^4];
            return name;
        }
    }

    /// <summary>
    /// PriorityLevel を ProcessPriorityClass に変換
    /// </summary>
    public static ProcessPriorityClass ToProcessPriorityClass(PriorityLevel level) => level switch
    {
        PriorityLevel.Idle => ProcessPriorityClass.Idle,
        PriorityLevel.BelowNormal => ProcessPriorityClass.BelowNormal,
        PriorityLevel.Normal => ProcessPriorityClass.Normal,
        PriorityLevel.AboveNormal => ProcessPriorityClass.AboveNormal,
        PriorityLevel.High => ProcessPriorityClass.High,
        PriorityLevel.Realtime => ProcessPriorityClass.RealTime,
        _ => ProcessPriorityClass.Normal
    };

    /// <summary>
    /// ルールが有効かどうか（ProcessName が空でない且つ IsEnabled）
    /// </summary>
    [JsonIgnore]
    public bool IsValid => !string.IsNullOrWhiteSpace(ProcessName);
}

File: src/cpp/models/ProcessPriorityRule.cpp
#include "ProcessPriorityRule.h"

#include "../win32/UnicodeUtil.h"

#include <algorithm>

namespace imeindicator::models {

const char* priorityLevelToString(PriorityLevel level) noexcept
{
    switch (level) {
        case PriorityLevel::Idle:        return "Idle";
        case PriorityLevel::BelowNormal: return "BelowNormal";
        case PriorityLevel::Normal:      return "Normal";
        case PriorityLevel::AboveNormal: return "AboveNormal";
        case PriorityLevel::High:        return "High";
        case PriorityLevel::Realtime:    return "Realtime";
    }
    return "Normal";
}

bool tryParsePriorityLevel(std::string_view s, PriorityLevel& out) noexcept
{
    if (s == "Idle")        { out = PriorityLevel::Idle;        return true; }
    if (s == "BelowNormal") { out = PriorityLevel::BelowNormal; return true; }
    if (s == "Normal")      { out = PriorityLevel::Normal;      return true; }
    if (s == "AboveNormal") { out = PriorityLevel::AboveNormal; return true; }
    if (s == "High")        { out = PriorityLevel::High;        return true; }
    if (s == "Realtime")    { out = PriorityLevel::Realtime;    return true; }
    return false;
}

DWORD priorityLevelToProcessPriorityClass(PriorityLevel level) noexcept
{
    switch (level) {
        case PriorityLevel::Idle:        return IDLE_PRIORITY_CLASS;
        case PriorityLevel::BelowNormal: return BELOW_NORMAL_PRIORITY_CLASS;
        case PriorityLevel::Normal:      return NORMAL_PRIORITY_CLASS;
        case PriorityLevel::AboveNormal: return ABOVE_NORMAL_PRIORITY_CLASS;
        case PriorityLevel::High:        return HIGH_PRIORITY_CLASS;
        case PriorityLevel::Realtime:    return REALTIME_PRIORITY_CLASS;
    }
    return NORMAL_PRIORITY_CLASS;
}

std::wstring ProcessPriorityRule::normalizedProcessName() const
{
    return win32::stripExeAndTrim(processName);
}

bool ProcessPriorityRule::isValid() const
{
    return !win32::trim(processName).empty();
}

int ProcessPriorityRule::validatedMaxBackoffExponent() const
{
    return std::clamp(maxBackoffExponent, 0, 10);
}

void to_json(nlohmann::json& j, const ProcessPriorityRule& r)
{
    j = nlohmann::json{
        {"processName",        win32::wideToUtf8(r.processName)},
        {"targetPriority",     priorityLevelToString(r.targetPriority)},
        {"maxBackoffExponent", r.maxBackoffExponent},
        {"isEnabled",          r.isEnabled},
        {"useECoreOnly",       r.useECoreOnly}
    };
}

void from_json(const nlohmann::json& j, ProcessPriorityRule& r)
{
    if (auto it = j.find("processName"); it != j.end() && it->is_string()) {
        r.processName = win32::utf8ToWide(it->get<std::string>());
    } else {
        r.processName.clear();
    }

    PriorityLevel parsedLevel = PriorityLevel::Normal;
    if (auto itStr = j.find("targetPriority");
        itStr != j.end() && itStr->is_string()) {
        if (!tryParsePriorityLevel(itStr->get<std::string>(), parsedLevel)) {
            parsedLevel = PriorityLevel::Normal;
        }
    } else if (auto itNum = j.find("targetPriority");
               itNum != j.end() && itNum->is_number_integer()) {
        // 数値で書かれていた場合のフォールバック
        const int v = itNum->get<int>();
        if (v >= 0 && v <= 5) parsedLevel = static_cast<PriorityLevel>(v);
    }
    r.targetPriority = parsedLevel;

    r.maxBackoffExponent = std::clamp(j.value("maxBackoffExponent", 6), 0, 10);
    r.isEnabled    = j.value("isEnabled",    true);
    r.useECoreOnly = j.value("useECoreOnly", false);
}

} // namespace imeindicator::models
